### PR TITLE
IGNITE-16043 added check for not null reader in BinaryObjectBuilderImpl

### PR DESCRIPTION
**Pull request by andreybell91**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/binary/builder/BinaryObjectBuilderImpl.java [hunk 1]
```diff
@@ -549,12 +549,14 @@ private void ensureReadCacheInit() {
...
                 return null;
         }
-        else {
+        else if (reader != null) {
             ensureReadCacheInit();
 
             int fldId = ctx.fieldId(typeId, name);
 
             val = readCache.get(fldId);
+        } else {
```
Comment (tledkov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/binary/builder/BinaryObjectBuilderImpl.java:558]: Please fix style

File: modules/core/src/test/java/org/apache/ignite/internal/binary/BinaryObjectBuilderAdditionalSelfTest.java [hunk 2]
```diff
@@ -1974,6 +1974,40 @@ public void testBuilderReusage() throws Exception {
...
+    @Test
+    public void testGetFieldForNotAssignedValInEmptyBuilder() {
+        BinaryObjectBuilder builder = binaries().builder("SomeType")
+                .setField("w", "wewe");
+
+        String fieldValue = builder.getField("field");
+        String wValue = builder.getField("w");
+
+        assertNull(fieldValue);
+        assertEquals("wewe", wValue);
```
Comment (tledkov-gridgain) [inline modules/core/src/test/java/org/apache/ignite/internal/binary/BinaryObjectBuilderAdditionalSelfTest.java:1989]:
```suggestion
        assertNull(builder.getField("field"));
        assertEquals("wewe", builder.getField("w"));
```

File: modules/core/src/test/java/org/apache/ignite/internal/binary/BinaryObjectBuilderAdditionalSelfTest.java [hunk 3]
```diff
@@ -1974,6 +1974,40 @@ public void testBuilderReusage() throws Exception {
...
+        BinaryObjectBuilderImpl builder = wrap(testObjectContainer);
+        builder.setField("w", "wewe");
+
+        String fieldValue = builder.getField("field");
+        String wValue = builder.getField("w");
+        String fooValue = builder.getField("foo");
+
+        assertNull(fieldValue);
+        assertEquals("wewe", wValue);
+        assertEquals("binaryCachedValue", fooValue);
```
Comment (tledkov-gridgain) [inline modules/core/src/test/java/org/apache/ignite/internal/binary/BinaryObjectBuilderAdditionalSelfTest.java:2008]:
```suggestion
        assertNull(builder.getField("field"));
        assertEquals("wewe", builder.getField("w"));
        assertEquals("binaryCachedValue", builder.getField("foo"));
```

File: modules/core/src/test/java/org/apache/ignite/internal/binary/BinaryObjectBuilderAdditionalSelfTest.java [hunk 4]
```diff
@@ -1974,6 +1974,40 @@ public void testBuilderReusage() throws Exception {
         assertEquals("val2", builder.build().field("f1"));
     }
 
+    /**
+     * @throws Exception If fails
+     */
+    @Test
+    public void testGetFieldForNotAssignedValInEmptyBuilder() {
```
Comment (tledkov-gridgain) [inline modules/core/src/test/java/org/apache/ignite/internal/binary/BinaryObjectBuilderAdditionalSelfTest.java:1981]: Please fix style: abbreviation shouldn't be used at method name.
suggest: `testGetFieldForNotAssignedValInEmptyBuilder` -> `testGetFieldForNotAssignedValueInEmptyBuilder` or  `testGetNotAssignedFieldInEmptyBuilder`

File: modules/core/src/main/java/org/apache/ignite/internal/binary/builder/BinaryObjectBuilderImpl.java (hunk 1 above)
Comment (andreybell91) [reply modules/core/src/main/java/org/apache/ignite/internal/binary/builder/BinaryObjectBuilderImpl.java]: Fixed

File: modules/core/src/test/java/org/apache/ignite/internal/binary/BinaryObjectBuilderAdditionalSelfTest.java (hunk 2 above)
Comment (andreybell91) [reply modules/core/src/test/java/org/apache/ignite/internal/binary/BinaryObjectBuilderAdditionalSelfTest.java]: Fixed

File: modules/core/src/test/java/org/apache/ignite/internal/binary/BinaryObjectBuilderAdditionalSelfTest.java (hunk 3 above)
Comment (andreybell91) [reply modules/core/src/test/java/org/apache/ignite/internal/binary/BinaryObjectBuilderAdditionalSelfTest.java]: Fixed

File: modules/core/src/test/java/org/apache/ignite/internal/binary/BinaryObjectBuilderAdditionalSelfTest.java (hunk 4 above)
Comment (andreybell91) [reply modules/core/src/test/java/org/apache/ignite/internal/binary/BinaryObjectBuilderAdditionalSelfTest.java]: Fixed

**Pull request merged**